File: evd_unity_sim_app/Assets/Application/_Scripts/evd_script (environment-simplified)/data/Region.cs
using System.Collections;
using System.Collections.Generic;


namespace EvD
{
    namespace Data
    {

        [System.Serializable]
        public class Region : Pose
        {
            /*
             * Class Constants
             */

            public static const double DEFAULT_ORIENTATION_LIMIT = 1.0;

            /*
             * Private Members
             */
            
            private bool _freeOrientation;
            private double _uncertaintyOrientationLimit;
            private Orientation _uncertaintyOrientationAltTarget = null;

            /* 
             * Constructors
             */

            public Region(Position centerPosition = null, Orientation centerOrientation = null, bool freeOrientation = true,
                          double uncertaintyOrientationLimit = DEFAULT_ORIENTATION_LIMIT, Orientation uncertaintyOrientationAltTarget = null,
                          string type = "", string name = "", string uuid = null, Node parent = null, bool appendType = true)
            : base(centerPosition,orientcenterOrientationation, appendType ? "region." + type : type, name,uuid,parent,appendType)
            {
                this.freeOrientation = freeOrientation;
                this.uncertaintyOrientationLimit = uncertaintyOrientationLimit;
                this.uncertaintyOrientationAltTarget = uncertaintyOrientationAltTarget;
            }

            public new static Region FromDct(Dictionary<string,object> dct)
            {
                return new Region(
                    centerPosition: Position.FromDict((Dictionary<string,object>)dct["center_position"]),
                    centerOrientation: Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]),
                    freeOrientation: (bool)dct["free_orientation"],
                    uncertaintyOrientationLimit: (double)dct["uncertainty_orientation_limit"],
                    uncertaintyOrientationAltTarget: Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]),
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }

            public override Dictionary<string,object> ToDict() 
            {
                var dct = base.ToDict();
                dct.Remove("position");
                dct.Remove("orientation");
                dct.Add("center_position",position.ToDict());
                dct.Add("center_orientation",orientation.ToDict());
                dct.Add("free_orientation",freeOrientation);
                dct.Add("uncertainty_orientation_limit",uncertaintyOrientationLimit);
                dct.Add("uncertainty_orientation_alt_target",uncertaintyOrientationAltTarget.ToDct());
                return dct;
            }

            /*
            * Accessors and Modifiers
            */

            public Position centerPosition
            {
                get
                {
                    return this.position;
                }

                set
                {
                    this.position = value;
                    UpdatedAttribute("center_position","set");
                }
            }

            public Orientation centerOrientation
            {
                get
                {
                    return this.orientation;
                }

                set
                {
                    this.orientation = value;
                    UpdatedAttribute("center_orientation","set");
                }
            }

            public bool freeOrientation
            {
                get
                {
                    return _freeOrientation;
                }

                set
                {
                    if (_freeOrientation != value) 
                    {
                        _freeOrientation = value;
                        UpdatedAttribute("free_orientation","set");
                    }
                }
            }

            public double uncertaintyOrientationLimit
            {
                get
                {
                    return _uncertaintyOrientationLimit;
                }

                set
                {
                    if (_uncertaintyOrientationLimit != value) 
                    {
                        _uncertaintyOrientationLimit = value;
                        UpdatedAttribute("uncertainty_orientation_limit","set");
                    }
                }
            }

            public Orientation uncertaintyOrientationAltTarget
            {
                get 
                {
                    return _uncertaintyOrientationAltTarget;
                }

                set
                {
                    if (_uncertaintyOrientationAltTarget != value) 
                    {
                        if (_uncertaintyOrientationAltTarget != null)
                        {
                            _uncertaintyOrientationAltTarget.RemoveFromCache();
                        }

                        _uncertaintyOrientationAltTarget = value;
                        if (_uncertaintyOrientationAltTarget != null) 
                        {
                            _uncertaintyOrientationAltTarget.parent = this;
                        }
                        
                        UpdatedAttribute("uncertainty_orientation_alt_target","set");
                    }
                }
            }

            public override void Set(Dictionary<string, object> dct)
            {
                if (dct.ContainsKey("center_position"))
                {
                    centerPosition = Position.FromDict((Dictionary<string,object>)dct["center_position"]);
                }

                if (dct.ContainsKey("center_orientation"))
                {
                    centerOrientation = Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]);
                }

                if (dct.ContainsKey("free_orientation"))
                {
                    freeOrientation = (bool)dct["free_orientation"];
                }

                if (dct.ContainsKey("uncertainty_orientation_limit"))
                {
                    uncertaintyOrientationLimit = (double)dct["uncertainty_orientation_limit"];
                }

                if (dct.ContainsKey("uncertainty_orientation_alt_target"))
                {
                    uncertaintyOrientationAltTarget = Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]);
                }

                base.Set(dct);
            }

            /*
             * Cache Methods
             */

            public override void RemoveFromCache()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.RemoveFromCache();
                }
                
                base.RemoveFromCache();
            }

            public override void AddToCache()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.AddToCache();
                }

                base.AddToCache();
            }

            /*
             * Update Methods
             */

            public override void LateConstructUpdate()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.LateConstructUpdate();
                }

                base.LateConstructUpdate();
            }

            public override void DeepUpdate()
            {
                if (uncertaintyOrientationAltTarget != null)
                {
                    uncertaintyOrientationAltTarget.DeepUpdate();
                }

                base.DeepUpdate();

                UpdatedAttribute("center_position","update");
                UpdatedAttribute("center_orientation","update");
                UpdatedAttribute("free_orientation","update");
                UpdatedAttribute("uncertainty_orientation_limit","update");
                UpdatedAttribute("uncertainty_orientation_alt_target","update");
            }

            public override void ShallowUpdate()
            {
                base.ShallowUpdate();

                UpdatedAttribute("center_position","update");
                UpdatedAttribute("center_orientation","update");
                UpdatedAttribute("free_orientation","update");
                UpdatedAttribute("uncertainty_orientation_limit","update");
                UpdatedAttribute("uncertainty_orientation_alt_target","update");
            }
        }

        [System.Serializable]
        public class CubeRegion : Region
        {
            /*
            * Private Members
            */

            private double _uncertaintyX;
            private double _uncertaintyY;
            private double _uncertaintyZ;

            /*
            * Constructors
            */

            public CubeRegion(Position centerPosition = null, Orientation centerOrientation = null, double uncertaintyX = 0.1,
                              double uncertaintyY = 0.1, double uncertaintyZ = 0.1, bool freeOrientation = true, 
                              double uncertaintyOrientationLimit = DEFAULT_ORIENTATION_LIMIT, Orientation uncertaintyOrientationAltTarget = null,
                              string type = "", string name = "", string uuid = null, Node parent = null, bool appendType = true)
            : base(centerPosition, orientcenterOrientationation, freeOrientation, uncertaintyOrientationLimit, uncertaintyOrientationAltTarget, 
                   appendType ? "cube-region." + type : type, name,uuid,parent,appendType)
            {
                this.uncertaintyX = uncertaintyX;
                this.uncertaintyY = uncertaintyY;
                this.uncertaintyZ = uncertaintyZ;
            }

            public new static CubeRegion FromDict(Dictionary<string,object> dct)
            {
                return new CubeRegion(
                    centerPosition: Position.FromDict((Dictionary<string,object>)dct["center_position"]),
                    centerOrientation: Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]),
                    uncertaintyX: (double)dct["uncertainty_x"],
                    uncertaintyY: (double)dct["uncertainty_y"],
                    uncertaintyZ: (double)dct["uncertainty_z"],
                    freeOrientation: (bool)dct["free_orientation"],
                    uncertaintyOrientationLimit: (double)dct["uncertainty_orientation_limit"],
                    uncertaintyOrientationAltTarget: Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]),
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }

            public override Dictionary<string, object> ToDict()
            {
                var dct = base.ToDict();
                dct.Add("uncertainty_x",uncertaintyX);
                dct.Add("uncertainty_y",uncertaintyY);
                dct.Add("uncertainty_z",uncertaintyZ);
                return dct;
            }

            /*
            * Accessors and Modifiers
            */

            public double uncertaintyX
            {
                get
                {
                    return _uncertaintyX;
                }

                set
                {
                    if (_uncertaintyX != value)
                    {
                        _uncertaintyX = value;
                        UpdatedAttribute("uncertainty_x","set");
                    }
                }
            }

            public double uncertaintyY
            {
                get
                {
                    return _uncertaintyY;
                }

                set
                {
                    if (_uncertaintyY != value)
                    {
                        _uncertaintyY = value;
                        UpdatedAttribute("uncertainty_y","set");
                    }
                }
            }

            public double uncertaintyZ
            {
                get
                {
                    return _uncertaintyZ;
                }

                set
                {
                    if (_uncertaintyZ != value)
                    {
                        _uncertaintyZ = value;
                        UpdatedAttribute("uncertainty_z","set");
                    }
                }
            }

            public override void Set(Dictionary<string, object> dct)
            {
                if (dct.ContainsKey("uncertainty_x"))
                {
                    uncertaintyX = (double)dct["uncertainty_x"];
                }

                if (dct.ContainsKey("uncertainty_y"))
                {
                    uncertaintyY = (double)dct["uncertainty_y"];
                }

                if (dct.ContainsKey("uncertainty_z"))
                {  
                    uncertaintyZ = (double)dct["uncertainty_z"];
                }

                base.Set(dct);
            }

            /*
            * Update Methods
            */

            public override void DeepUpdate()
            {

                base.DeepUpdate();

                UpdatedAttribute("uncertainty_x", "update");
                UpdatedAttribute("uncertainty_y", "update");
                UpdatedAttribute("uncertainty_z", "update");
            }

            public override void ShallowUpdate()
            {
                base.ShallowUpdate();

                UpdatedAttribute("uncertainty_x", "update");
                UpdatedAttribute("uncertainty_y", "update");
                UpdatedAttribute("uncertainty_z", "update");
            }
        }


        [System.Serializable]
        public class SphereRegion : Region
        {
            /*
            * Private Members
            */

            private double _uncertaintyRadius;

            /*
            * Constructors
            */

            public SphereRegion(Position centerPosition = null, Orientation centerOrientation = null, double uncertaintyRadius = 0.1,
                              double uncertaintyOrientationLimit = DEFAULT_ORIENTATION_LIMIT, Orientation uncertaintyOrientationAltTarget = null,
                              string type = "", string name = "", string uuid = null, Node parent = null, bool appendType = true)
            : base(centerPosition, orientcenterOrientationation, freeOrientation, uncertaintyOrientationLimit, uncertaintyOrientationAltTarget, 
                   appendType ? "sphere-region." + type : type, name,uuid,parent,appendType)
            {
                this.uncertaintyRadius = uncertaintyRadius;
            }

            public new static SphereRegion FromDict(Dictionary<string,object> dct)
            {
                return new SphereRegion(
                    centerPosition: Position.FromDict((Dictionary<string,object>)dct["center_position"]),
                    centerOrientation: Orientation.FromDict((Dictionary<string,object>)dct["center_orientation"]),
                    uncertaintyRadius: (double)dct["uncertainty_radius"],
                    freeOrientation: (bool)dct["free_orientation"],
                    uncertaintyOrientationLimit: (double)dct["uncertainty_orientation_limit"],
                    uncertaintyOrientationAltTarget: Orientation.FromDict((Dictionary<string,object>)dct["uncertainty_orientation_alt_target"]),
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }

            public override Dictionary<string, object> ToDict()
            {
                var dct = base.ToDict();
                dct.Add("uncertainty_radius",uncertaintyRadius);
                return dct;
            }

            /*
            * Accessors and Modifiers
            */

            public double uncertaintyRadius
            {
                get
                {
                    return _uncertaintyRadius;
                }

                set
                {
                    if (_uncertaintyRadius != value)
                    {
                        _uncertaintyRadius = value;
                        UpdatedAttribute("uncertainty_radius","set");
                    }
                }
            }

            public override void Set(Dictionary<string, object> dct)
            {
                if (dct.ContainsKey("uncertainty_radius"))
                {
                    uncertaintyRadius = (double)dct["uncertainty_radius"];
                }

                base.Set(dct);
            }

            /*
            * Update Methods
            */

            public override void DeepUpdate()
            {

                base.DeepUpdate();

                UpdatedAttribute("uncertainty_radius", "update");
            }

            public override void ShallowUpdate()
            {
                base.ShallowUpdate();

                UpdatedAttribute("uncertainty_radius", "update");
            }
        }
    }
}

File: evd_ros_backend/evd_ros_core/src/evd_script/data/regions/region.py
'''
Regions are just poses with uncertain orientation (and potentially position in child-types)

NOTE, right now regions are on parallel development to waypoints/locations and these two do
not mingle when discussing robots. Future work should address this.
'''

from ...node_parser import NodeParser
from ..geometry import Pose, Position, Orientation
from ...visualizable import VisualizeMarker

class Region(Pose, VisualizeMarker):

    '''
    Class Constants
    '''

    DEFAULT_ORIENTATION_LIMIT = 1

    '''
    Data structure methods
    '''

    @classmethod
    def type_string(cls, trailing_delim=True):
        return 'region' + ('.' if trailing_delim else '')

    @classmethod
    def full_type_string(cls):
        return Pose.full_type_string() + cls.type_string()

    def __init__(self, center_position=None, center_orientation=None, free_orientation=True,
                 uncertainty_orientation_limit=1, uncertainty_orientation_alt_target=None,
                 type='', name='', uuid=None, parent=None, append_type=True, editable=True,
                 deleteable=True, description=''):
        self._free_ort = None
        self._uncert_ort_lim = None
        self._uncert_ort_target = None

        super(Region,self).__init__(
            position=center_position,
            orientation=center_orientation,
            type=Region.type_string() + type if append_type else type,
            name=name,
            uuid=uuid,
            parent=parent,
            append_type=append_type,
            editable=editable,
            deleteable=deleteable,
            description=description)

        self.free_orientation = free_orientation
        self.uncertainty_orientation_limit = uncertainty_orientation_limit if not free_orientation else None
        self.uncertainty_orientation_alt_target = uncertainty_orientation_alt_target if not free_orientation else None

    def to_dct(self):
        msg = super(Region,self).to_dct()

        del msg['position']         # overriding with center_position
        del msg['orientation']      # overriding with center_orientation

        msg.update({
            'center_position': self.position.to_dct(),
            'center_orientation': self.orientation.to_dct(),
            'free_orientation': self.free_orientation,
            'uncertainty_orientation_limit': self.uncertainty_orientation_limit,
            'uncertainty_orientation_alt_target': self.uncertainty_orientation_alt_target.to_dct()
        })
        return msg

    @classmethod
    def from_dct(cls, dct):
        return cls(
            center_position=NodeParser(dct['center_position'], enforce_types=[Position.type_string(trailing_delim=False)]),
            center_orientation=NodeParser(dct['center_orientation'], enforce_types=[Orientation.type_string(trailing_delim=False)]),
            free_orientation=dct['free_orientation'],
            uncertainty_orientation_limit=dct['uncertainty_orientation_limit'],
            uncertainty_orientation_alt_target=NodeParser(dct['uncertainty_orientation_alt_target'], enforce_types=[Orientation.type_string(trailing_delim=False)]),
            type=dct['type'] if 'type' in dct.keys() else '',
            append_type=not 'type' in dct.keys(),
            uuid=dct['uuid'] if 'uuid' in dct.keys() else None,
            name=dct['name'] if 'name' in dct.keys() else '')

    def to_ros_marker(self, frame_id='app', id=0):
        return None # Cannot visualize an un-poisitionally-bounded marker

    '''
    Data accessor/modifier methods
    '''

    @property
    def center_position(self):
        return self.position

    @center_position.setter
    def center_position(self, value):
        self.position = value
        self.updated_attribute("center_position","set")

    @property
    def center_orientation(self):
        return self.orientation

    @center_orientation.setter
    def center_orientation(self, value):
        self.orientation = value
        self.updated_attribute("center_orientation","set")

    @property
    def free_orientation(self):
        return self._free_ort

    @free_orientation.setter
    def free_orientation(self, value):
        # Note will destory previous uncertainty values

        if self._free_ort != value:
            if type(value) != type(True):
                raise Exception("Value must be a boolean")

            self._free_ort = value
            if value:
                self.uncertainty_orientation_limit = None
                self.uncertainty_orientation_alt_target = None
            else:
                self.uncertainty_orientation_limit = self.DEFAULT_ORIENTATION_LIMIT
                self.uncertainty_orientation_alt_target = None

            self.updated_attribute('free_orientation','set')

    @property
    def uncertainty_orientation_limit(self):
        return self._uncert_ort_lim

    @uncertainty_orientation_limit.setter
    def uncertainty_orientation_limit(self, value):
        if self._uncert_ort_lim != value:
            if self._free_ort:
                raise Exception('Orientation is set as free, change free_orientation to False first!')

            if value < 0:
                raise Exception('Value must be positive or zero')

            self._uncert_ort_lim = value
            self.updated_attribute('uncertainty_orientation_limit','set')

    @property
    def uncertainty_orientation_alt_target(self):
        return self._uncert_ort_target

    @uncertainty_orientation_alt_target.setter
    def uncertainty_orientation_alt_target(self, value):
        if self._uncert_ort_target != value:
            if self._free_ort:
                raise Exception('Orientation is set as free, change free_orientation to False first!')

            if self._uncert_ort_target != None:
                self._uncert_ort_target.remove_from_cache()

            self._uncert_ort_target = value
            if self._uncert_ort_target != None:
                self._uncert_ort_target.parent = self

            self.updated_attribute('uncertainty_orientation_alt_target','set')

    def set(self, dct):

        if 'center_position' in dct.keys():
            self.center_position = NodeParser(dct["center_position"], enforce_types=[Position.type_string(trailing_delim=False)])

        if 'center_orientation' in dct.keys():
            self.center_orientation = NodeParser(dct["center_orientation"], enforce_types=[Orientation.type_string(trailing_delim=False)])

        if 'free_orientation' in dct.keys():
            self.free_orientation = dct['free_orientation']

        if 'uncertainty_orientation_limit' in dct.keys():
            self.uncertainty_orientation_limit = dct['uncertainty_orientation_limit']

        if 'uncertainty_orientation_alt_target' in dct.keys():
            self.uncertainty_orientation_alt_target = NodeParser(dct['uncertainty_orientation_alt_target'], enforce_types=[Orientation.type_string(trailing_delim=False)]) if dct['uncertainty_orientation_alt_target'] != None else None

        super(Region,self).set(dct)

    def check_if_pose_within_uncertainty(self, pose):
        if self.free_orientation:
            return True
        else:
            target = self.uncertainty_orientation_alt_target if self.uncertainty_orientation_alt_target != None else self.orientation

            dif = Orientation.difference(pose.orientation,target)

            return dif <= self.uncertainty_orientation_limit

    '''
    Cache Methods
    '''

    def remove_from_cache(self):

        if self.uncertainty_orientation_alt_target != None:
            self.uncertainty_orientation_alt_target.remove_from_cache()

        super(Region,self).remove_from_cache()

    def add_to_cache(self):

        if self.uncertainty_orientation_alt_target != None:
            self.uncertainty_orientation_alt_target.add_to_cache()

        super(Region,self).add_to_cache()

    '''
    Update Methods
    '''

    def late_construct_update(self):

        if self.uncertainty_orientation_alt_target != None:
            self.uncertainty_orientation_alt_target.late_construct_update()

        super(Region,self).late_construct_update()

    def deep_update(self):

        if self.uncertainty_orientation_alt_target != None:
            self.uncertainty_orientation_alt_target.deep_update()

        super(Region,self).deep_update()

        self.updated_attribute('center_position', 'update')
        self.updated_attribute('center_orientation', 'update')
        self.updated_attribute('free_orientation','update')
        self.updated_attribute('uncertainty_orientation_limit','update')
        self.updated_attribute('uncertainty_orientation_alt_target','update')

    def shallow_update(self):
        super(Region,self).shallow_update()

        self.updated_attribute('center_position', 'update')
        self.updated_attribute('center_orientation', 'update')
        self.updated_attribute('free_orientation','update')
        self.updated_attribute('uncertainty_orientation_limit','update')
        self.updated_attribute('uncertainty_orientation_alt_target','update')
